app/controllers/accounts_controller.rb

```diff
@@ -26,4 +26,11 @@ def create
   helper_method memoize def account
     Account.visible_for(current_user).find(ps.fetch(:id))
   end
+
+  helper_method memoize def read_only?(id)
+    return false if current_user.account.id == account.parent.id ||
+                    AccountShare.accepted.for(current_user).pluck(:account_id).include?(id)
+
+    AccountShare.accepted.for_public.pluck(:account_id).include?(id)
+  end
```

Comment:
Instead of query DB for each account we can cache the list of public shared accounts and it will be much faster. for now it will be too slow if there are many accounts.

```ruby
memoize def shared_accounts
  Set.new(AccountShare.accepted.for_public.pluck(:account_id))
end

helper_method memoize def public_share?(account)
  shared_accounts.include?(account.id)
end
```

And now use `public_share?(account)` everywhere in the views. Note, that we don't need to check for other conditions here because the buttons will be accessible otherwise.